### PR TITLE
[java] Improve module grammar

### PR DESCRIPTION
**Pull request by oowekyala**
I believe this is the last grammar change needed for PMD 7. We can proceed with cleaning up the AST as in #3831 

## Related issues

- An item of #2701

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleDirective.java [hunk 1]
```diff
@@ -4,44 +4,47 @@
...
+    @NoAttribute
+    public DirectiveType getType() {
+        return type;
     }
 
-    void setRequiresModifier(RequiresModifier requiresModifier) {
-        this.requiresModifier = requiresModifier;
+    /**
+     * Kind of a module directive. Specific kinds are represented by
+     * specific subclasses.
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleDirective.java:44]: Do we still need this enum then when we have the subclasses?

File: pmd-java/etc/grammar/Java.jjt [hunk 2]
```diff
@@ -2734,40 +2739,34 @@ void DefaultValue():
...
-  | ( LOOKAHEAD({isKeyword("provides")}) <IDENTIFIER> { jjtThis.setType(ASTModuleDirective.DirectiveType.PROVIDES); } Name() LOOKAHEAD({isKeyword("with")}) <IDENTIFIER> Name() ("," Name() )* ";" )
+    ( LOOKAHEAD({isKeyword("requires")}) <IDENTIFIER>
+      [
+        LOOKAHEAD({isKeyword("transitive")}) <IDENTIFIER> { jjtThis.setTransitive(); }
+      |                                      "static"     { jjtThis.setStatic(); }
+      ]
+      ModuleName() ";" ) #ModuleRequiresDirective
+  | ( LOOKAHEAD({isKeyword("exports")})  <IDENTIFIER> packageName=VoidNameNoLookahead() {jjtThis.setPackageName(packageName);} [ LOOKAHEAD({isKeyword("to")}) <IDENTIFIER> ModuleName() ("," ModuleName())* ] ";" ) #ModuleExportsDirective
+  | ( LOOKAHEAD({isKeyword("opens")})    <IDENTIFIER> packageName=VoidNameNoLookahead() {jjtThis.setPackageName(packageName);} [ LOOKAHEAD({isKeyword("to")}) <IDENTIFIER> ModuleName() ("," ModuleName())* ] ";" ) #ModuleOpensDirective
+  | ( LOOKAHEAD({isKeyword("uses")})     <IDENTIFIER> packageName=VoidNameNoLookahead() {jjtThis.setPackageName(packageName);} ";" ) #ModuleUsesDirective
```
Comment (adangel) [inline pmd-java/etc/grammar/Java.jjt:2762]: uses should use also a ClassName like provides. In JLS it's a "TypeName" (https://docs.oracle.com/javase/specs/jls/se17/html/jls-7.html#jls-7.7)

```suggestion
  | ( LOOKAHEAD({isKeyword("uses")})     <IDENTIFIER> ClassName() ";" ) #ModuleUsesDirective
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleUsesDirective.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * A "uses" directive of a {@linkplain ASTModuleDeclaration module declaration}.
+ *
+ * <pre class="grammar">
+ *
+ * ModuleUsesDirective ::= "uses" &lt;PACKAGE_NAME&gt; ";"
+ *
+ * </pre>
+ */
+public final class ASTModuleUsesDirective extends AbstractPackageNameModuleDirective {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleUsesDirective.java:16]:
```suggestion
public final class ASTModuleUsesDirective extends ASTModuleDirective {
```

Probably a helper method like in provides makes sense:

```java
    public ASTClassOrInterfaceType getService() {
        return firstChild(ASTClassOrInterfaceType.class);
    }
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleProvidesDirective.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    ASTModuleProvidesDirective(int id) {
+        super(id, DirectiveType.PROVIDES);
+    }
+
+    @Override
+    protected <P, R> R acceptVisitor(JavaVisitor<? super P, ? extends R> visitor, P data) {
+        return visitor.visit(this, data);
+    }
+
+    public ASTClassOrInterfaceType getProvidedInterface() {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleProvidesDirective.java:32]: Maybe call this "getService()"? - the JLS talks about "services" and "service providers"

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleProvidesDirective.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        return firstChild(ASTClassOrInterfaceType.class);
+    }
+
+    public NodeStream<ASTClassOrInterfaceType> getImplementations() {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleProvidesDirective.java:36]: Maybe "getServiceProviders()"?

**Review by adangel — COMMENTED** [review]
Thanks for updating this!

I've added a new section in <https://github.com/pmd/pmd/wiki/Java_clean_changes#module-declarations> - updated already with ModuleUsesDirective using a ClassOrInterfaceType as child node instead of package name.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleDirective.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleDirective.java:44]: Removing it

**Comment by oowekyala** [comment]
Thanks for updating the wiki and spotting that mistake about the uses directive!

**Review by adangel — APPROVED** [review]

**Pull request merged**